### PR TITLE
[framework] sorting of validators

### PR DESCRIPTION
**Pull request by davidiw**
* Add a Move comparator that uses bcs, so we can sort validator addresses
* Add a sorting of validators

Learn more here: https://github.com/aptos-labs/aptos-core/issues/1362

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/Comparator.move [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    #[test]
+    public fun test_strings() {
+        use Std::ASCII;
+
+        let value0 = ASCII::string(b"alpha");
+        let value1 = ASCII::string(b"beta");
+        let value2 = ASCII::string(b"betaa");
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/Comparator.move:71]: nit: it made me wonder why there isn't `assert!(is_equal(&compare(&value1, &value1)), x);`

File: aptos-move/framework/aptos-framework/sources/Comparator.move [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        assert!(is_smaller_than(&compare(&value1, &value0)), 2);
+        assert!(is_greater_than(&compare(&value0, &value1)), 3);
+        assert!(is_smaller_than(&compare(&value0, &value2)), 4);
+        assert!(is_smaller_than(&compare(&value1, &value2)), 5);
+        assert!(is_greater_than(&compare(&value2, &value0)), 6);
+        assert!(is_greater_than(&compare(&value2, &value1)), 7);
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/Comparator.move:77]: -_- the ordering is a myth

File: aptos-move/framework/aptos-framework/sources/Comparator.move [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    #[test]
+    public fun test_u128() {
+        let value0: u128 = 5;
+        let value1: u128 = 152;
+        let value2: u128 = 511; // 0x1ff
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/Comparator.move:87]: nit: value2?

File: aptos-move/framework/aptos-framework/sources/Comparator.move [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        let value0_0 = Vector::empty();
+        Vector::push_back(&mut value0_0, 10);
+        Vector::push_back(&mut value0_0, 9);
+        Vector::push_back(&mut value0_0, 5);
+
+        let value0_1 = Vector::empty();
+        Vector::push_back(&mut value0_0, 10);
+        Vector::push_back(&mut value0_0, 9);
+        Vector::push_back(&mut value0_0, 5);
+        Vector::push_back(&mut value0_0, 1);
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/Comparator.move:112]: do you mean value0_1?

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 2 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Comparator.move:77]: are you arguing that the ordering appears inconsistent?

beta < alpha < betaa

because it is (string length, string contents)
bcs is great, we need traits or interfaces....

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 1 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Comparator.move:71]: I don't understand. Maybe you wanna see value2?

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 3 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Comparator.move:87]: for completeness!

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 4 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Comparator.move:112]: muahaha, I was wondering why my brain couldn't figure out comparing base and other_0....

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 1 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Comparator.move:71]: 😞

File: aptos-move/framework/aptos-framework/sources/Comparator.move (hunk 2 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Comparator.move]: I mean, I'd assert in this order (or something else that I can follow better):
0 < 1
1 > 0
0 < 2
2 > 0
1 < 2
2 > 1

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Comparator.move]: That was my second guess, so I already restructured it ;)

File: aptos-move/framework/aptos-framework/sources/configs/Stake.move [hunk 5]
```diff
@@ -456,9 +457,33 @@ module AptosFramework::Stake {
             };
             i = i + 1;
         };
+
+        sort_validators(&mut active_validators);
         validator_set.active_validators = active_validators;
     }
 
+    fun sort_validators(validators: &mut vector<ValidatorInfo>) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/Stake.move:465]: I thought you wanted to do this in a native function? This is fine too as I'm not sure if there's a big performance gain from that.

File: aptos-move/framework/aptos-framework/sources/configs/Stake.move [hunk 6]
```diff
@@ -456,9 +457,33 @@ module AptosFramework::Stake {
...
+        let length = Vector::length(validators);
+        if (length == 0) {
+            return
+        };
+
+        let ordered = false;
+        while (!ordered) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/Stake.move:472]: What is this? Bubble sort? 😵‍💫 Can this tank performance here if the validator set is large enough (thousands)?

File: config/src/generator.rs [hunk 7]
```diff
@@ -50,6 +50,12 @@ pub fn validator_swarm(
         network.seeds = seeds.clone();
     }
 
+    nodes.sort_by(|a, b| {
```
Comment (movekevin) [inline config/src/generator.rs:53]: Why is this needed?

File: types/src/on_chain_config/validator_set.rs [hunk 8]
```diff
@@ -64,6 +67,22 @@ impl ValidatorSet {
     pub fn empty() -> Self {
         ValidatorSet::new(Vec::new())
     }
+
+    fn ordered_validators(payload: &[ValidatorInfo]) -> bool {
```
Comment (movekevin) [inline types/src/on_chain_config/validator_set.rs:71]: Is there a function you can call directly (like Vector::sort) on payload here to sort and then compare the before and after?

File: types/src/proof/definition.rs [hunk 9]
```diff
@@ -516,7 +516,7 @@ where
 
             // Next we take two children at a time and compute their parents.
             let mut children_iter = children_iter.as_slice().chunks_exact(2);
-            while let Some(chunk) = children_iter.next() {
+            for chunk in children_iter.by_ref() {
```
Comment (movekevin) [inline types/src/proof/definition.rs:519]: Is this just a minor unrelated refactor?

File: aptos-move/framework/aptos-framework/sources/configs/Stake.move (hunk 6 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/configs/Stake.move:472]: Maybe, but remember this is used once a day or a few times a day. This is something that can easily be worked on in the background. Other searching schemes involve more complexity and testing -- especially in Move. I'm not keen to do that given the constraints -- the lack of callbacks / closures.

Bubble sort ftw.

File: config/src/generator.rs (hunk 7 above)
Comment (davidiw) [reply config/src/generator.rs:53]: I wanted the debug assert to ensure that they were sorted by move, but in some cases we generate a validator set independent of genesis -- and then that debug assert will fail.

File: types/src/on_chain_config/validator_set.rs (hunk 8 above)
Comment (davidiw) [reply types/src/on_chain_config/validator_set.rs:71]: I mean, I guess we could, but why, when this works?

File: types/src/proof/definition.rs (hunk 9 above)
Comment (davidiw) [reply types/src/proof/definition.rs:519]: clippy picked up on it and was giving me warnings.

File: aptos-move/framework/aptos-framework/sources/configs/Stake.move (hunk 6 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/configs/Stake.move:472]: That makes sense! I think this is a new achievement unlocked though - implementing bubble sort on-chain 😆

File: types/src/on_chain_config/validator_set.rs (hunk 8 above)
Comment (movekevin) [reply types/src/on_chain_config/validator_set.rs:71]: Just nit I guess - I was wondering if this can be done in 2 lines instead of 13 lines.

**Review by movekevin — APPROVED** [review]

File: types/src/on_chain_config/validator_set.rs (hunk 8 above)
Comment (davidiw) [reply types/src/on_chain_config/validator_set.rs:71]: muahahaha, I like my silly approach... sort may have been better ... maybe?

fwiw, this is compiled away in release

**Review by davidiw — COMMENTED** [review]
/land

**Review by zekun000 — COMMENTED** [review]
we need to deal with pending_inactive validators too, since it's used in https://github.com/aptos-labs/aptos-core/blob/9ca9bc40250165b0973e1a248465135bba929d27/types/src/on_chain_config/validator_set.rs#L64

we probably need to duplicate them in the active_validators and remove them in the on_new_epoch function. (it's currently moved from active to pending_inactive)

**Comment by davidiw** [comment]
> we need to deal with pending_inactive validators too, since it's used in
> 
> https://github.com/aptos-labs/aptos-core/blob/9ca9bc40250165b0973e1a248465135bba929d27/types/src/on_chain_config/validator_set.rs#L64
> 
> we probably need to duplicate them in the active_validators and remove them in the on_new_epoch function. (it's currently moved from active to pending_inactive)

Good call and my debug assert should be in that function

File: aptos-move/framework/aptos-framework/sources/configs/Stake.move [hunk 10]
```diff
@@ -696,6 +692,93 @@ module AptosFramework::Stake {
...
+        on_new_epoch();
+        assert!(validator_index(v1_addr) == 0, 10);
+        assert!(validator_index(v2_addr) == 1, 11);
+        assert!(validator_index(v3_addr) == 2, 12);
+        assert!(validator_index(v4_addr) == 3, 13);
+        assert!(validator_index(v5_addr) == 4, 14);
+    }
+
+    #[test_only]
+    public(script) fun register_mint_stake(
```
Comment (Manahip) [inline aptos-move/framework/aptos-framework/sources/configs/Stake.move:764]: Should we change this to `public` instead of `public(script)`? @davidiw my transpiler is grudgy about having a struct as parameter in a `public(script)`
